pkg/controller/machineset.go

```diff
@@ -677,6 +678,64 @@ func (c *controller) terminateMachines(inactiveMachines []*v1alpha1.Machine, mac
 	return nil
 }
 
+func (c *controller) terminateStaleMachines(activeMachines, inactiveMachines []*v1alpha1.Machine, machineSet *v1alpha1.MachineSet) error {
+	var (
+		wg                      sync.WaitGroup
+		numOfMachinesToDelete   int
+		numOfRunningMachines    int
+		numOfMachineSetReplicas = int(machineSet.Spec.Replicas)
+		numOfActiveMachines     = len(activeMachines)
+		numOfInactiveMachines   = len(inactiveMachines)
+		errCh                   = make(chan error, numOfInactiveMachines)
+	)
+	defer close(errCh)
+
+	if numOfInactiveMachines == 0 {
+		return nil
+	}
+
+	for _, activeMachine := range activeMachines {
+		if IsMachineRunning(activeMachine) {
+			numOfRunningMachines++
+		}
+	}
+
+	if numOfRunningMachines == 0 || numOfMachineSetReplicas == 0 || c.safetyOptions.FailedMachineDeletionRatio == 1 || c.safetyOptions.FailedMachineDeletionRatio == 0 {
+		numOfMachinesToDelete = numOfInactiveMachines
+	} else {
+		numOfMachinesToDelete = int(math.Round(float64(numOfInactiveMachines) * c.safetyOptions.FailedMachineDeletionRatio))
+		if numOfMachinesToDelete == 0 {
+			numOfMachinesToDelete = 1
+		}
+
+		pendingMachines := numOfActiveMachines - numOfRunningMachines
+
+		if pendingMachines < numOfMachinesToDelete {
+			numOfMachinesToDelete -= pendingMachines
+		} else {
+			klog.Infof("Skipping terminateStaleMachines iteration on MachineSet %q/%q, previous batch is not Running yet", machineSet.Namespace, machineSet.Name)
+			numOfMachinesToDelete = 0
+		}
+	}
+
+	wg.Add(numOfMachinesToDelete)
+	for i := 0; i < numOfMachinesToDelete; i++ {
```

Comment:
a validation `if FailedMachineDeletionRatio <= 1` might be useful to make sure `numOfMachinesToDelete < len(inactiveMachines)`

Reply:
Done in the previously not used Validate() method.